src/DIRAC/Core/Tornado/Client/private/TornadoBaseClient.py

```diff
@@ -221,6 +232,21 @@ def __discoverCredentialsToUse(self):
       else:
         self.kwargs[self.KW_SKIP_CA_CHECK] = skipCACheck()
 
+    # Use tokens?
+    if self.KW_USE_ACCESS_TOKEN in self.kwargs:
+      self.__useAccessToken = self.kwargs[self.KW_USE_ACCESS_TOKEN]
+    else:
+      self.__useAccessToken = gConfig.getValue("/DIRAC/Security/UseTokens", "false").lower() in ("y", "yes", "true")
+      self.kwargs[self.KW_USE_ACCESS_TOKEN] = self.__useAccessToken
+      if 'DIRAC_USE_ACCESS_TOKEN' in os.environ:
+        self.__useAccessToken = os.environ.get('DIRAC_USE_ACCESS_TOKEN', 'false').lower() in ['yes', 'true']
```

Comment:
I'm struggling to follow this. Why is `self.kwargs[self.KW_USE_ACCESS_TOKEN]` set if the CS has tokens enabled but not if the environment variable is set?

Reply:
Good question, `self.kwargs[self.KW_USE_ACCESS_TOKEN]` should depend on the final value of `self.__useAccessToken`. Will be fixed to:
```
    # Use tokens?
    if self.KW_USE_ACCESS_TOKEN in self.kwargs:
      self.__useAccessToken = self.kwargs[self.KW_USE_ACCESS_TOKEN]
    else:
      self.__useAccessToken = gConfig.getValue("/DIRAC/Security/UseTokens", "false").lower() in ("y", "yes", "true")
      if 'DIRAC_USE_ACCESS_TOKEN' in os.environ:
        self.__useAccessToken = os.environ.get('DIRAC_USE_ACCESS_TOKEN', 'false').lower() in ['yes', 'true']
      self.kwargs[self.KW_USE_ACCESS_TOKEN] = self.__useAccessToken
```

Reply:
Sounds good but how about removing a level of indentation:
```python
    if self.KW_USE_ACCESS_TOKEN in self.kwargs:
      self.__useAccessToken = self.kwargs[self.KW_USE_ACCESS_TOKEN]
    elif 'DIRAC_USE_ACCESS_TOKEN' in os.environ:
      self.__useAccessToken = os.environ.get('DIRAC_USE_ACCESS_TOKEN', 'false').lower() in ['yes', 'true']
    else:
      self.__useAccessToken = gConfig.getValue("/DIRAC/Security/UseTokens", "false").lower() in ("y", "yes", "true")
    self.kwargs[self.KW_USE_ACCESS_TOKEN] = self.__useAccessToken
```

Reply:
Yes, it looks better. I can be also without two lines:
```
    self.__useAccessToken = self.kwargs.get(self.KW_USE_ACCESS_TOKEN,
                                            os.environ.get('DIRAC_USE_ACCESS_TOKEN', 
                                                           gConfig.getValue("/DIRAC/Security/UseTokens")))
    self.__useAccessToken = (self.__useAccessToken or 'false').lower() in ("y", "yes", "true")
    self.kwargs[self.KW_USE_ACCESS_TOKEN] = self.__useAccessToken
```

Reply:
I think it's more readable to keep it as a simple `if...elif...else`.

In fact I suspect there is a bug in your shorted version as `self.__useAccessToken` should be a boolean if it's in `self.kwargs` which then won't have a `.lower()` method. Plus `self.kwargs[self.KW_USE_ACCESS_TOKEN] == False` will be ignored.

Reply:
Yep agree, your con-fu is stronger )